Add options to set the fft size and to normalize the ifft output; calculate the fft by default; and improve the documentation and the tests

AngelEzquerra opened the pull request:
This PR does a few small things:
- It improves the fft function documentation
- It adds a normalize argument to the fft and dft functions. Normalization is disabled by default but if enabled it will normalize the ifft otuput (i.e. it only applies if inverse is true)
- It makes inverse=false the default (i.e. the fft function calculates the fft by default)
- It adds an `n` argument to make it possible to choose the FFT size (it still uses the input length by default).
- It improves the tests (which could fail if the folds were negative)

AngelEzquerra commented:
I wonder if there is a more optimum way to implement the "Add `n` argument to select the FFT size" commit. That is, to only create a copy of the input when needed.

arnetheduck commented on src/fftr.nim:
```diff
@@ -323,25 +323,105 @@ func process*(
+    for i in 0..<result.len:
+      result[i] = result[i] / complexInputLen
+
+func fft*(input: openArray[Complex64], inverse: bool = false, normalize: bool = false, n = 0): seq[Complex64] =
```
regarding normalization, there are two kinds - `1/n` and `1/sqrt(n)` depending on what the numbers will be used for - see https://dsp.stackexchange.com/a/63006 and https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fft.html#scipy.fft.fft

regarding `n`, this feels like outside of the domain of the fft function itself - it's a `setLen` away for anyone that wants to do fft this particular way (though they probably don't)

AngelEzquerra replied:
I knew that scipy's fft had 3 normalization options. If you think that makes sense I can support them all by passing a norm argument that would be an enum.

Regarding n, I think it would be a good idea to support it as well, following scpy's API as much as possible.

AngelEzquerra replied:
I've added `norm` arguments matching SCIPY's options plus a "disabled" option to completely disable the normalization. I've made the default `backwards` to mach SCIPY.

arnetheduck commented on src/fftr.nim:
```diff
@@ -323,25 +323,105 @@ func process*(
+    input.toSeq & newSeq[Complex64](size - input.len)
+  elif size < input.len:
+    # Crop the input as much as needed
+    input[0..<size]
```
if `n` was pursued, this would introduce an unnecessary copy (that could be avoided with `toOpenArray`)

AngelEzquerra replied:
You are right of course. I even mentioned in my first comment on this PR :)
I will look into it.

AngelEzquerra replied:
I've improved the code to avoid unnecessary copies.

arnetheduck commented on src/fftr.nim:
```diff
@@ -323,25 +323,105 @@ func process*(
+
+  let input = if size > input.len:
+    # Extend the input with zeros as much as needed
+    input.toSeq & newSeq[Complex64](size - input.len)
```
if `n` was pursued, `setLen` here would involve fewer copies and zeroings

AngelEzquerra replied:
I've improved the code to avoid unnecessary copies.

arnetheduck commented on src/fftr.nim:
```diff
@@ -323,25 +323,105 @@ func process*(
+    # Crop the input as much as needed
+    input[0..<size]
+  else:
+    input.toSeq
```
and another copy here - adding these copies up would significantly slow down the implementation - you should see this if you run the benchmark before and after

AngelEzquerra replied:
FWIW, I ran the benchmarks and I did not notice any significant differences, but I agree that unnecessary copies should be avoided.

AngelEzquerra replied:
I've improved the code to avoid unnecessary copies.

AngelEzquerra committed:
Add documentation to the main fft function

AngelEzquerra committed:
Calculate the fft by default (if inverse is not provided)

AngelEzquerra committed:
add `normalize` argument to the fft function

AngelEzquerra committed:
add abs to the test checks to avoid false passes when the folds are negative

AngelEzquerra committed:
Reduce the amount of copies when using the `n` argument of the fft procedure

AngelEzquerra committed:
Add `n` argument to select the FFT size

The argument name has been chosen to emulate the numpy.fft API.

AngelEzquerra committed:
Add separate tests for FFT with specific sizes larger, equal and smaller than the input.

AngelEzquerra committed:
Swap the positions of the norm and n arguments of the fft function

This makes the fft and ifft functions use the same argument order, as well as making this library match scpy's fft argument order.

AngelEzquerra committed:
Replace the fft normalization parameter with an enum norm parameter (to match the scpy normalization feature API).

The default normalization is set to `backward` to match the scpy behavior.

This commit also adds tests for the different normalization modes.

AngelEzquerra committed:
Add examples of the new arguments to the README
